enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,281 @@
+to specify and access a set of shared CPUs, whether those shared CPUs are shared among containers in the 
+same pod or containers in different ones.
+
+By giving the users the flexibility to use exclusive CPUs only when it is an absolute necessity, we help them utilize
+their existing compute resources in an optimal manner.
+
+### User Stories
+
+* As a developer who's building DPDK application,
+I want to run my application in a container,
+giving full exclusive access to cpu cores for the busy-loops polling threads,
+but having the light-weight tasks like configuration processing and log printing not taking full cores,
+in order to improve the application density and make more efficient use of the CPUs.
+For that, my application container needs access to both shared and exclusive CPUs.
+
+  
+### Goals 
+* Increase pod density, power efficiency, cost reduction and sustainability by optimizing CPUs resource management. 
+* Provide a special-purpose at a runtime-level solution for containers in a pod to have the ability
+to request and subsequently be allocated both exclusive and shared CPUs.
+* Enable the telco-specific use case with a self-contained change,
+    minimizing the changes to the platform
+
+### Non-Goals
+* Introducing a generic mechanism in the platform that does involve Kubelet and pod spec changes.
+
+## Proposal
+
+Container processes are restricted
+to given [cpuset](https://man7.org/linux/man-pages/man7/cpuset.7.html) by their corresponding cgroup definition.
+Essentially, the kernel enforces the processes of container to run only at the CPU (core) ids specified under its
+cgroup.
+
+The mixed-cpu-node-plugin is a  runtime-agnostic, hybrid (composed of both [NRI-plugin](https://github.com/containerd/nri) and
+[device-plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)), 
+[mixed-cpu-node-plugin](https://github.com/openshift-kni/mixed-cpu-node-plugin) 
+that communicates with the container-runtime
+at various lifecycle stages of container e.g. CreateContainer, UpdateContainer, etc.,
+and modifies the container's cgroup to include a set of shared CPUs.
+
+The shared CPUs are part of the [reserved-cpus](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list) CPUs pool.
+The reason to utilize the reserved CPUs is because of CPU manager which is not aware of the CPUs
+lying in this pool, so it doesn't undo or changes the allocation logic performed by the plugin.
+Therefore, the plugin assigns the shared CPUs to containers without racing/conflict
+the CPU manager behavior. 
+More about this decision at the [Alternative](mixed-cpu-node-plugin.md#alternatives) section 
+
+The node-plugin populates a special device named `openshift.io/shared-cpus` to provide way for pods to request
+for this special type of CPUs.
+There's no meaning to the value/number of `openshift.io/shared-cpus` devices that the pod requests.
```

Comment:
Is this request going to be used by the scheduler to place the pod on a node that has these resources? That should be discussed here.

Reply:
This is part of the scheduler behavior, but since it's a dummy device it would be populated to all nodes.
 The actual reason for having it as a resource described at lines 110 -114:
_"The reason for specifying a device, and not, for example, an annotation,
is because when application pod requests for a device, the scheduler makes the pod pending till device up and running.
This gives the node plugin room for setup and not being depended on pod admission order.
So only when the node plugin finishes the registration process with NRI and device plugin, the scheduler
admits the application pod."_

Reply:
> I'm curious about the case where the shared-cpus are not configured on a node in its performance profile, so that node would not publish the openshift.io/shared-cpus resource

Yes. If shared-cpus are not configured, NTO won't deploy the node-plugin that responsible  to populate the `openshift.io/shared-cpus` devices.
>In that case, will the scheduler avoid putting pods with requests for that resource on those nodes?

If the pods asks for a device, the scheduler schedules the pod only to nodes that have this certain device. if non of the nodes have an available device, the pod keep pending. this is a native behavior of the scheduler.

> Or is the requirement that all nodes need to have shared-cpus configured - and if so, how would that be enforced?

No it's not a requirement.